functional-tests/appwrapper_reconciler_test.go

```diff
@@ -0,0 +1,250 @@
+package functional_tests
+
+import (
+	"context"
+	"go/build"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	gstruct "github.com/onsi/gomega/gstruct"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	. "github.com/project-codeflare/codeflare-common/support"
+	"github.com/project-codeflare/instascale/controllers"
+	"github.com/project-codeflare/instascale/pkg/config"
+	mcadv1beta1 "github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/apis/controller/v1beta1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	log "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	cfg       *rest.Config
+	k8sClient client.Client // You'll be using this client in your tests.
+	testEnv   *envtest.Environment
+	err       error
+)
+
+func logger() {
+	// Get log file path from environment variable or use a default path
+	logFilePath := os.Getenv("LOG_FILE_PATH")
+	if logFilePath == "" {
+		logFilePath = "./functional-test-logfile.log"
+	}
+	// Create a log file
+	logFile, err := os.Create(logFilePath)
+	if err != nil {
+		log.Log.Error(err, "Error creating log file: %v", err)
+	}
+	// Configure zap logger to write to the log file
+	logger := zap.New(zap.WriteTo(logFile), zap.UseDevMode(true))
+
+	// Set the logger for controller-runtime
+	ctrl.SetLogger(logger)
+
+	// This line prevents controller-runtime from complaining about log.SetLogger never being called
+	log.SetLogger(logger)
+}
+
+func startEnvTest(t *testing.T) {
+	// to redirect all functional test related logs to separate logfile ~ default (in functional-tests directory), can be changed using environment variable LOG_FILE_PATH=/path_to_logfile
+	logger()
+
+	//specify testEnv configuration
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "project-codeflare", "multi-cluster-app-dispatcher@v1.39.0", "config", "crd", "bases"),
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "openshift", "api@v0.0.0-20230213134911-7ba313770556", "machine", "v1beta1"),
+		},
+		ErrorIfCRDPathMissing: true,
+	}
+	test := WithConfig(t, testEnv.Config)
+	cfg, err = testEnv.Start()
+	test.Expect(err).NotTo(HaveOccurred())
+
+	t.Cleanup(func() {
+		teardownTestEnv(t, testEnv)
+	})
+
+	establishClient(t, testEnv) // setup client required for test to interact with kubernetes machine API
+}
+
+func establishClient(t *testing.T, testEnv *envtest.Environment) {
+	// Create a test instance with the provided configuration.
+	test := WithConfig(t, cfg)
```

Comment:
If this function is called from test function then you can directly pass `Test` struct and avoid creating it here.